### DISCUSSION TITLE
Lack of matches due to rate limiting?

### DISCUSSION THREAD

**Discussion by ghbrown** [beetbox/beets#4691] [Q&A]
When running `beet import -s` on a folder with many songs I often get `No matching recordings found` for Discogs. I suspect this can happen due to rate limiting (which my verbose logs sometimes show) in addition to an actual lack of a match.

However, when I enter essentially the same query on the Discogs website I get the correct results. To be clear, these queries are so simple I think it is almost impossible that special characters, etc. are causing an a true lack of matches, and what I suspect is happening is the following:

```
$ beet import -s dir/
*beets makes a lot of API calls for tracks*
*Discogs rate limits me, causing one or more failed requests*
*beets attaches these failed requests to the track and stores this result*
*I get prompted with results for a few successful queries*
*I make it to prompt for failed queries: No matching recordings found*
*rate limit wears off, so I can use manual search mode*
```

Is this possible? What else could be causing this, are there any simple ways to say for sure?

Also, are there any flags or options I can set to limit how many cached matches I have at once? I would prefer to suffer a bit of latency if it means I don't need to do all these searches manually.

**Comment by sampsyo** [comment]
Hmm; I suppose it's plausible! It would be really great to confirm in two possible ways (ideally both):

* A verbose log: `beet import -vv -s dir/` should reveal more information.
* Trying it again specifically on the songs that failed. If importing the same songs that couldn't be matched before _alone_ suddenly starts working, that would certainly be evidence for the rate-limiting theory!

**Reply by ghbrown** [reply]
I copied quite a few tracks already tagged by `beets` into a new directory and ran an import (into an empty database). I specifically tried to choose well known tracks with a minimum of special characters that I remember being able to find on Discogs yesterday (automatically or manually).

At least on the user side, it feels like the same thing is happening again, and from what I can tell [the logs](https://github.com/beetbox/beets/files/10894178/beets_log.txt) seem to confirm my original theory. In the logs: quite a few requests are successfully made, then rate limit messages appear, then I navigate a few prompts and suddenly I'm getting `No matching recordings found.` but I can still make a successful manual search.

A good example of this would be the tracks by Dazz Band. If you search for these in the logs you'll first see a rate limited http request, then eventually my activities in the prompt where I put in a manual search that is successful.
(Note that the successful request has a hyphen in it. I hope this is an artifact of the log message, but this is the only difference between the queries I can see (and hyphens can apparently cause issues for Discogs), though obviously it has nothing to do with rate limiting.

**Reply by sampsyo** [reply]
Curious! We removed rate limiting on our side in https://github.com/beetbox/beets/pull/4118 because it seemed that the underlying library had added its own rate limiting. Maybe it would be a good idea to check the version of the Discogs library you have (perhaps `pip freeze | grep discogs` could help) and compare that against the version where they added proper rate limiting? Or just make sure it's upgraded to the latest version!

**Reply by ghbrown** [reply]
Yep, that was it, thanks. I installed the `discogs-client` package instead of `python3-discogs-client` while in a rush to set up (and based off of an error message instead of the plugin docs).

Would implementing a check for this and generating a warning be possible (I'm happy to do it if so)? I saw in the PR you linked that it was supposed to have a requirement for a recent version of `python3-discogs-client`, but I was still able to up and running with the wrong package.

**Reply by sampsyo** [reply]
Hmm; good question… a warning surely would be useful! I'm just not 100% sure how to go about detecting which library/version is installed. Maybe the module has a `__version__` member we can check?

**Reply by ghbrown** [reply]
There is a `__version__` variable in [`__init__.py`](https://github.com/joalla/discogs_client/blob/master/discogs_client/__init__.py) if that would work.

**Reply by sampsyo** [reply]
That seems like the thing! Any chance you'd feel comfortable putting together a pull request to check that version?

**Reply by ghbrown** [reply]
I'd be happy to make a PR for this, I'm just not sure which method might be preferable:
- It should be easy enough to check the value of `__version__` after import somewhere. However, I'm worried that since both modules are named `discogs_client` that checking the version this way won't be sufficient to distinguish the two in the case that the PYPI package `discogs-client` has some version larger than `2.3.10`.
- I found [`importlib.metadata`](https://docs.python.org/3/library/importlib.metadata.html#module-importlib.metadata) (built in after Python 3.8), which takes a package name and returns a version string. So it would be able to differentiate between `python3-discogs-client` and `discogs-client`. Not sure if using this would hurt any efforts to support old versions, etc.
- I could also check for existence of different functions, etc. between the two modules and use that to deduce which package is present.

**Reply by sampsyo** [reply]
I could be misremembering this, but I believe the reason the new library exists is because `discogs-client` is officially abandoned—i.e., it will not be getting new versions. So I think we can be confident that there will be never be a `discogs-client` with a version greater than `2.3.10`.
